source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+    {
+      string[] mixture;
+      foreach (member; __traits(allMembers, Struct)) {
+        if (__traits(getOverloads, Struct, member).length == 0) {
+          //pragma(msg, member);
+          mixture ~= member ~ ":" ~ (member == Name ? "value" : member);
+        }
+      }
+      return mixture.join(",\n");
+    }(),
+    "};"
+  );
+}
+
+
+immutable struct Function
+{
+
+  /**
+     A string that evaluates to a function symbol.
+   */
+
+  string localSymbol;
+
+  /**
+     Function name. Initial value: `__traits(identifier, fun)`.
+   */
+
+  string name;
+
+  /**
+     Return a new Function object with the name attribute set to value.
+  */
+
+  Function setName(string value)
+  {
+    mixin replaceAttribute!"name";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    pure @nogc int answer();
+    mixin(refract!(answer, "answer").setName("ultimateAnswer").mixture);
+    static assert(
+      __traits(getAttributes, ultimateAnswer) ==
+      __traits(getAttributes, answer));
+  }
+
+  /**
+     Index of function in a set of overloads. Valid only if Function represents
+     a function (not a function pointer), and was refracted from a module.
+   */
+
+  uint index;
+
+  /**
+     Return a new Function object with the index attribute set to value.
+  */
+
+  Function setIndex(uint value)
+  {
+    mixin replaceAttribute!"index";
+    return copy;
+  }
+
+  /**
+     Return a string representing the parameter.
+   */
+
+  /**
+     Return type. Initial value: `std.traits.ReturnType!fun`.
+   */
+
+  string returnType;
+
+  /**
+     Return a new Function object with the returnType attribute set to value.
+  */
+
+  Function setReturnType(string value)
+  {
+    mixin replaceAttribute!"returnType";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    pure int answer() { return 42; }
+    mixin(
+      refract!(answer, "answer").setName("realAnswer")
+      .setReturnType("real")
+      .mixture);
+    static assert(is(typeof(realAnswer()) == real));
+    static assert(functionAttributes!realAnswer & FunctionAttribute.pure_);
+  }
+
+  /**
+     Function parameters. Initial value: from the refracted function.
+   */
+
+  Parameter[] parameters;
+
+  /**
+     Return a new Function object with the parameters attribute set to value.
+  */
+
+  Function setParameters(immutable(Parameter)[] value)
+  {
+    mixin replaceAttribute!"parameters";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    int answer();
+    mixin(
+      refract!(answer, "answer").setName("answerQuestion")
+      .setParameters([ Parameter().setName("question").setType("string")])
+      .mixture);
+    int control(string);
+    static assert(is(Parameters!answerQuestion == Parameters!control));
+  }
+
+  /**
+     Return a new Function object with a parameter attribute containing the
+     same parameters, with new parameters inserted parameters at the specified
+     index.
+  */
+
+  Function insertParameters(uint index, immutable(Parameter)[] inserted...)
```

Comment:
Suggestion to name `insertParametersAt`?

Reply:
> Suggestion to name `insertParametersAt`?

I don't know why but I dislike plurals inside identifiers. To me it's clear that any function with ïnsert" in its name takes a position, while "add" or "append" do not. Also, `insertParametersAt` what? Which leads us to `insertParametersAtPosition`. 

But I'm not stubborn about this, let me know if you really prefer with At.

Reply:
Insert can be used for quite a bit more than ints. Eg. on `sets` it is the element. On a linked list it could be the element to insert after or insert before, etc.

Though I do not have a strong opinion for at really. I am trying to think of the usability on this function. It seems out of place almost since it doesn't use the same `with...` prefix like others. So I think I would actually propose `withParametersAt`. 

WDYT?